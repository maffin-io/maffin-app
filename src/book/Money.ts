import * as currencies from '@dinero.js/currencies';
import type { Currency as DineroCurrency } from 'dinero.js';
import * as djs from 'dinero.js';

import { toAmountWithScale, moneyToString } from '@/helpers/number';

export default class Money {
  private _raw: djs.Dinero<number>;

  /**
   * A wrapper class for dinero object to reduce verbosity and overhead. The constructor
   * checks for how many decimals the amount has and calculates the scale automatically.
   *
   * @param amount - The amount to be considered in the money object. Can be a float
   * @param currency - The currency
   * @param scale - Number representing custom scale. Default is 0, if passed differently, that one
   * is taken into account
   */
  constructor(amount: number, currency: string, scale?: number) {
    const dineroCurrency = toCurrency(currency);

    if (scale === undefined) {
      const { amount: newAmount, scale: newScale } = toAmountWithScale(amount);
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount: newAmount,
        scale: newScale,
      });
    } else {
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount,
        scale,
      });
    }
  }

  get currency(): string {
    return djs.toSnapshot(this._raw).currency.code;
  }

  get raw(): djs.Dinero<number> {
    return this._raw;
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   */
  toString(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => `${value} ${currency.code}`,
    );
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   *
   * The result is a localized string with the currency as a symbol
   */
  format(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => moneyToString(Number(value), currency.code),
    );
  }

  toNumber(): number {
    return parseFloat(djs.toDecimal(this._raw));
  }

  isNegative(): boolean {
    return djs.isNegative(this._raw);
  }

  isZero(): boolean {
    return djs.isZero(this._raw);
  }

  /**
   * Adds b to the current object and returns a new one with the result
   *
   * @param b - Money to be added
   * @returns - Money with result
   */
  add(b: Money): Money {
    const result = djs.toSnapshot(djs.add(b.raw, this.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Subtracts b to the current object and returns a new one with the result
   *
   * @param b - Money to be subtracted
   * @returns - Money with result
   */
  subtract(b: Money): Money {
    const result = djs.toSnapshot(djs.subtract(this.raw, b.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Multiplies b to the current object and returns a new one with the result
   * Initially, we try to multiply without transforming b but sometimes dinerojs
   * returns that it is an invalid amount (I have no idea why). If it fails, then
   * we transform b to `amountWithScale` and multiply with that.
   *
   * @param b - amount to be multiplied to in scale 0
   * @returns - Money with result
   */
  multiply(b: number): Money {
    let result;

    try {
      result = djs.multiply(this.raw, b);
    } catch {
      result = djs.multiply(this.raw, toAmountWithScale(b));
    }

    const snapshot = djs.toSnapshot(result);
    return new Money(snapshot.amount, this.currency, snapshot.scale);
  }

  /**
   * Checks for equality with b by comparing amount and currency
   *
   * @returns - boolean
   */
  equals(b: Money): boolean {
    return djs.equal(this._raw, b._raw);
  }

  /**
   * Converts a Money object to the currency specified
   *
   * @param to - Money object containing the rate to
   */
  convert(
    to: string,
    rate: number,
  ): Money {
    const rates = {
      [to]: {
        ...toAmountWithScale(rate),
      },
    };
    const currency = toCurrency(to);

    try {
      // @ts-ignore
      const result = djs.toSnapshot(djs.convert(this.raw, currency, rates));
      return new Money(result.amount, currency.code, result.scale);
    } catch (e) {
      throw new Error(`Error converting ${this.format()} to ${to} with rates ${JSON.stringify(rates)}: ${e}`);
    }
  }
}

function toCurrency(currency: string): DineroCurrency<number> {
  // @ts-ignore
  let dineroCurrency = currencies[currency];
  if (!dineroCurrency) {
    dineroCurrency = {
      code: currency,
      base: 10,
      exponent: 2,
    };
  }
  return dineroCurrency;
}
