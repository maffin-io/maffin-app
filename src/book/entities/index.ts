export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
