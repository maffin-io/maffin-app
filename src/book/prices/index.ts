export { default as PriceDBMap } from './PriceDBMap';
