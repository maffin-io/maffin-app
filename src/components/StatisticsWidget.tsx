import React from 'react';

export type StatisticsWidgetProps = {
  title: string,
  stats: string,
  description: string,
  className?: string,
  statsTextClass?: string,
};

export default function StatisticsWidget({
  title,
  stats,
  description,
  className = '',
  statsTextClass = '',
}: StatisticsWidgetProps): JSX.Element {
  return (
    <div className={`${className} bg-gunmetal-700 rounded-sm p-6`}>
      <p>
        {title}
      </p>
      <p className={`${statsTextClass} text-2xl font-semibold my-3`}>{stats}</p>

      <span className="text-sm">
        {description}
      </span>
    </div>
  );
}
