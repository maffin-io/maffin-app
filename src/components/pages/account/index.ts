export { default as TotalLineChart } from './TotalLineChart';
export { default as SplitsHistogram } from './SplitsHistogram';
export { default as TransactionsTable } from './TransactionsTable';
export { default as Header } from './Header';
export { default as AccountInfo } from './AccountInfo';
export { default as InvestmentInfo } from './InvestmentInfo';
export { default as AssetInfo } from './AssetInfo';
