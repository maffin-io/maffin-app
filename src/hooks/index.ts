export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
