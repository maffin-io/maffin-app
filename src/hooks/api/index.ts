import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
