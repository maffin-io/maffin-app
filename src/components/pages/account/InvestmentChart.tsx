import React from 'react';
import { Chart, Filler } from 'chart.js';
import { DateTime, Interval } from 'luxon';

import { Price } from '@/book/entities';
import Line from '@/components/charts/Line';
import { moneyToString } from '@/helpers/number';
import type { Account } from '@/book/entities';
import { useInvestments, usePrices } from '@/hooks/api';
import Loading from '@/components/Loading';

Chart.register(Filler);

export type InvestmentChartProps = {
  account: Account,
};

export default function InvestmentChart({
  account,
}: InvestmentChartProps): JSX.Element {
  const { data: investments } = useInvestments(account.guid);
  let { data: prices } = usePrices(account.commodity.guid);

  const investment = investments?.[0];
  if (!investment || !prices) {
    return <Loading />;
  }

  const startDate = investment.account.splits[0].transaction.date;
  prices = prices || [];
  const currency = (prices.length && prices[0].currency.mnemonic) || '';

  const pricesData: { x: number, y: number }[] = [];
  prices.forEach(price => {
    pricesData.push({
      x: price.date.toMillis(),
      y: price.value,
    });
  });

  const numStocksData: { x: number, y: number }[] = [];
  const valueData: { x: number, y: number }[] = [];
  const interval = Interval.fromDateTimes(startDate, DateTime.now());
  const monthly = interval.splitBy({ month: 1 }).map(d => (d.start as DateTime).startOf('month'));
  monthly.forEach(x => {
    investment.processSplits(x);
    numStocksData.push({
      x: x.toMillis(),
      y: investment.quantity.toNumber(),
    });

    const price = findClosestPrice(x, prices || []);
    if (price) {
      valueData.push({
        x: x.toMillis(),
        y: investment.quantity.toNumber() * price.value,
      });
    }
  });

  return (
    <Line
      height="400px"
      data={{
        datasets: [
          {
            label: 'Num. stocks',
            data: numStocksData,
            yAxisID: 'yStocks',
            // @ts-ignore
            type: 'bar',
            borderColor: 'rgba(124, 58, 237, 1)',
            backgroundColor: 'rgba(124, 58, 237, 1)',
            order: 0,
          },
          {
            label: 'Price',
            data: pricesData,
            yAxisID: 'yPrice',
            pointStyle: 'circle',
            tension: 0.4,
            cubicInterpolationMode: 'monotone',
            fill: false,
            borderColor: 'rgba(234, 88, 12, 1)',
            backgroundColor: 'rgba(234, 88, 12, 1)',
            order: 1,
          },
          {
            label: 'Value',
            data: valueData,
            yAxisID: 'yValue',
            fill: false,
            pointRadius: 5,
            borderColor: 'rgba(22, 163, 74)',
            backgroundColor: 'rgba(22, 163, 74)',
            order: 2,
          },
        ],
      }}
      options={{
        maintainAspectRatio: false,
        plugins: {
          datalabels: {
            display: false,
          },
          legend: {
            display: true,
            onClick: () => {},
            labels: {
              pointStyle: 'circle',
              usePointStyle: true,
            },
          },
          tooltip: {
            backgroundColor: '#323b44',
            callbacks: {
              label: (item) => {
                if (item.datasetIndex !== 1) {
                  return `${moneyToString(Number(item.parsed.y), investment.account.commodity.mnemonic)}`;
                }

                return `${moneyToString(Number(item.parsed.y), currency)}`;
              },
            },
          },
        },
        scales: {
          x: {
            type: 'time',
            time: {
              unit: 'month',
              tooltipFormat: 'dd MMMM yyyy',
              displayFormats: {
                month: 'MMM-yy',
              },
            },
            grid: {
              offset: false,
            },
            border: {
              display: false,
            },
            min: startDate.startOf('month').toISODate() as string,
          },
          yStocks: {
            offset: true,
            beginAtZero: true,
            stackWeight: 2,
            stack: 'investment',
            grid: {
              display: false,
            },
            border: {
              display: false,
            },
            ticks: {
              color: 'rgba(124, 58, 237, 1)',
              callback: (value) => moneyToString(
                value as number,
                investment.account.commodity.mnemonic,
              ),
            },
          },
          yPrice: {
            offset: true,
            stackWeight: 4,
            stack: 'investment',
            border: {
              display: false,
            },
            ticks: {
              color: 'rgba(234, 88, 12, 0.7)',
              maxTicksLimit: 10,
              callback: (value) => moneyToString(value as number, currency),
            },
          },
          yValue: {
            offset: true,
            stackWeight: 4,
            stack: 'investment',
            border: {
              display: false,
            },
            ticks: {
              color: 'rgba(22, 163, 74)',
              maxTicksLimit: 10,
              callback: (value) => moneyToString(value as number, currency),
            },
          },
        },
      }}
    />
  );
}

function findClosestPrice(date: DateTime, prices: Price[]): Price {
  const possible = prices.filter(price => price.date <= date);
  return possible[possible.length - 1];
}
