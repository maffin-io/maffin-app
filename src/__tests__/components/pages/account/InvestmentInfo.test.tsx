import React from 'react';
import { render, screen } from '@testing-library/react';
import { DateTime, Interval } from 'luxon';
import type { DefinedUseQueryResult, UseQueryResult } from '@tanstack/react-query';

import { InvestmentInfo } from '@/components/pages/account';
import InvestmentChart from '@/components/pages/account/InvestmentChart';
import StatisticsWidget from '@/components/StatisticsWidget';
import type { Account, Commodity } from '@/book/entities';
import { Price } from '@/book/entities';
import Money from '@/book/Money';
import * as apiHook from '@/hooks/api';
import * as stateHooks from '@/hooks/state';
import { PriceDBMap } from '@/book/prices';
import type { InvestmentAccount } from '@/book/models';

jest.mock('@/components/pages/account/InvestmentChart', () => jest.fn(
  () => <div data-testid="InvestmentChart" />,
));

jest.mock('@/components/StatisticsWidget', () => jest.fn(
  () => <div data-testid="StatisticsWidget" />,
));

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

jest.mock('@/hooks/state', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/state'),
}));

jest.mock('@/components/Loading', () => jest.fn(
  () => <div data-testid="Loading" />,
));

describe('InvestmentInfo', () => {
  let eur: Commodity;
  let ticker: Commodity;

  beforeEach(() => {
    jest.spyOn(apiHook, 'usePrices').mockReturnValue({ data: undefined } as UseQueryResult<PriceDBMap>);
    jest.spyOn(apiHook, 'useInvestment').mockReturnValue({ data: undefined } as UseQueryResult<InvestmentAccount>);
    jest.spyOn(stateHooks, 'useInterval').mockReturnValue({ data: TEST_INTERVAL } as DefinedUseQueryResult<Interval>);

    jest.spyOn(Price, 'create').mockImplementation();

    eur = {
      guid: 'eur_guid',
      mnemonic: 'EUR',
      namespace: 'CURRENCY',
    } as Commodity;

    ticker = {
      guid: 'ticker_guid',
      mnemonic: 'TICKER',
      namespace: 'OTHER',
    } as Commodity;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('returns loading when no data', async () => {
    render(
      <InvestmentInfo
        account={
          {
            guid: 'guid',
            commodity: eur,
          } as Account
        }
      />,
    );

    await screen.findByTestId('Loading');
  });

  it('renders as expected with data', async () => {
    const account = {
      guid: 'guid',
      commodity: ticker,
    } as Account;
    jest.spyOn(apiHook, 'usePrices').mockReturnValue({
      data: new PriceDBMap([
        {
          date: DateTime.fromISO('2022-12-01'),
          value: 10,
          currency: eur,
          commodity: ticker,
        } as Price,
        {
          date: DateTime.fromISO('2023-01-01'),
          value: 15,
          currency: eur,
          commodity: ticker,
        } as Price,
      ]),
    } as UseQueryResult<PriceDBMap>);
    jest.spyOn(apiHook, 'useInvestment').mockReturnValue({
      data: {
        cost: new Money(100, 'EUR'),
        value: new Money(150, 'EUR'),
        unrealizedProfitAbs: new Money(50, 'EUR'),
        unrealizedProfitPct: 50,
        realizedProfit: new Money(10, 'EUR'),
        realizedProfitPct: 5,
        avgPrice: 10,
        quantity: new Money(10, 'TICKER'),
        account,
        currency: 'EUR',
        realizedDividends: new Money(20, 'EUR'),
      },
    } as UseQueryResult<InvestmentAccount>);

    const { container } = render(
      <InvestmentInfo account={account} />,
    );

    await screen.findByText('10 titles');
    await screen.findByText('€10.00');

    expect(InvestmentChart).toBeCalledWith(
      {
        account,
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'col-span-6',
        title: 'Closest price',
        statsTextClass: 'table-caption badge default',
        stats: '€15.00',
        description: 'on 1/1/2023',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        className: 'col-span-6',
        title: 'Value',
        statsTextClass: 'amount-positive',
        stats: '€150.00',
        description: 'from €100.00 invested',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      3,
      {
        className: 'col-span-6',
        title: 'Unrealized Profit',
        statsTextClass: 'amount-positive',
        stats: '50 %',
        description: expect.anything(),
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      4,
      {
        className: 'col-span-6',
        title: 'Realized Profit',
        statsTextClass: 'amount-positive',
        stats: '€10.00',
        description: '+ €20.00 from dividends',
      },
      {},
    );

    expect(container).toMatchSnapshot();
  });

  it('renders as expected when dividend with different currency', async () => {
    const account = {
      guid: 'guid',
      commodity: ticker,
    } as Account;
    jest.spyOn(apiHook, 'usePrices').mockReturnValue({
      data: new PriceDBMap([
        {
          date: DateTime.fromISO('2023-01-01'),
          value: 10,
          currency: eur,
          commodity: ticker,
        } as Price,
        {
          date: DateTime.fromISO('2023-02-01'),
          value: 15,
          currency: eur,
          commodity: ticker,
        } as Price,
      ]),
    } as UseQueryResult<PriceDBMap>);
    jest.spyOn(apiHook, 'useInvestment').mockReturnValue({
      data: {
        cost: new Money(100, 'EUR'),
        value: new Money(150, 'EUR'),
        unrealizedProfitAbs: new Money(50, 'EUR'),
        unrealizedProfitPct: 50,
        realizedProfit: new Money(10, 'EUR'),
        realizedProfitPct: 5,
        avgPrice: 10,
        quantity: new Money(10, 'TICKER'),
        account,
        currency: 'EUR',
        realizedDividends: new Money(20, 'USD'),
      },
    } as UseQueryResult<InvestmentAccount>);

    render(
      <InvestmentInfo account={account} />,
    );

    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      4,
      {
        className: 'col-span-6',
        title: 'Realized Profit',
        statsTextClass: 'amount-positive',
        stats: '€10.00',
        description: '+ $20.00 from dividends',
      },
      {},
    );
  });
});
