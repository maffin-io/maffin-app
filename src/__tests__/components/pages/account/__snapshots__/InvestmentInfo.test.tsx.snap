// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestmentInfo renders as expected with data 1`] = `
<div>
  <div
    class="grid grid-cols-12 items-stretch"
  >
    <div
      class="col-span-4"
    >
      <div
        class="card text-lg"
      >
        <span
          class="badge default text-xl font-semibold mx-1"
        >
          10 titles
        </span>
        <span>
          at an average price of
        </span>
        <span
          class="badge default text-xl font-semibold mx-1"
        >
          â‚¬10.00
        </span>
      </div>
      <div
        class="grid grid-cols-12"
      >
        <div
          data-testid="StatisticsWidget"
        />
        <div
          data-testid="StatisticsWidget"
        />
        <div
          data-testid="StatisticsWidget"
        />
        <div
          data-testid="StatisticsWidget"
        />
      </div>
    </div>
    <div
      class="card col-span-8"
    >
      <div
        data-testid="InvestmentChart"
      />
    </div>
  </div>
</div>
`;
